File: Tutorials/Codes/System_Dynamics_Course/Chapter15/Lesson4/Chapter15_Lesson4.m
% Chapter15_Lesson4.m
% Event handling and hybrid dynamics in MATLAB / Simulink-oriented style
% Example: Bouncing ball with ode45 Events + reset map loop

function Chapter15_Lesson4
    params.g = 9.81;
    params.e = 0.82;
    params.minBounceSpeed = 0.05;

    t0 = 0.0;
    tf = 8.0;
    y0 = [1.5; 0.0]; % [height; velocity]

    T_all = [];
    Y_all = [];
    impactTimes = [];

    opts = odeset('Events', @(t,y) groundEvent(t,y), ...
                  'RelTol', 1e-8, 'AbsTol', 1e-10, 'MaxStep', 0.03);

    tStart = t0;
    yStart = y0;

    while tStart < tf
        [T, Y, TE, YE, IE] = ode45(@(t,y) flow(t,y,params), [tStart tf], yStart, opts); %#ok<ASGLU>
        T_all = [T_all; T]; %#ok<AGROW>
        Y_all = [Y_all; Y]; %#ok<AGROW>

        if isempty(TE)
            break;
        end

        impactTimes(end+1,1) = TE(1); %#ok<AGROW>
        yMinus = YE(1,:).';

        if abs(yMinus(2)) < params.minBounceSpeed
            T_all = [T_all; TE(1)]; %#ok<AGROW>
            Y_all = [Y_all; 0.0, 0.0]; %#ok<AGROW>
            break;
        end

        yPlus = [0.0; -params.e * yMinus(2)];
        tStart = TE(1);
        yStart = yPlus + [0.0; 1e-12]; % avoid immediate retrigger
    end

    figure('Color', 'w');
    plot(T_all, Y_all(:,1), 'LineWidth', 1.5);
    grid on;
    xlabel('Time (s)');
    ylabel('Height (m)');
    title('Hybrid Simulation with Events (Bouncing Ball)');

    disp('Impact times:');
    disp(impactTimes);

    % Simulink note:
    % Use Integrator blocks with zero-crossing enabled, a Hit Crossing block
    % on h(t)=0, and a Triggered Subsystem / Stateflow chart to apply the reset
    % v^+ = -e v^- at each detected event.
end

function dydt = flow(~, y, params)
    dydt = [y(2); -params.g];
end

function [value, isterminal, direction] = groundEvent(~, y)
    value = y(1);      % event when height reaches zero
    isterminal = 1;    % stop integration and return control to script
    direction = -1;    % detect descending crossing only
end

File: Tutorials/Codes/System_Dynamics_Course/Chapter15/Lesson4/Chapter15_Lesson4.py
# Chapter15_Lesson4.py
"""
Event handling, switching, and hybrid continuous-discrete dynamics
Example: Bouncing ball with restitution using solve_ivp event detection.
"""

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt


g = 9.81           # gravitational acceleration (m/s^2)
e = 0.82           # coefficient of restitution
t0, tf = 0.0, 8.0  # simulation horizon
y0 = np.array([1.5, 0.0])  # [height, velocity]
min_bounce_speed = 0.05     # terminate when impacts become negligible


def flow(t, y):
    h, v = y
    return [v, -g]


def ground_event(t, y):
    # Event function g(y) = h. Trigger at h = 0 while descending.
    return y[0]


ground_event.terminal = True
ground_event.direction = -1.0


def jump_map(y_minus):
    h_minus, v_minus = y_minus
    # Reset state at impact: h^+ = 0, v^+ = -e v^-
    return np.array([0.0, -e * v_minus])


def simulate_hybrid_bouncing_ball():
    t_segments = []
    y_segments = []
    impact_times = []

    t_curr = t0
    y_curr = y0.copy()

    while t_curr < tf:
        sol = solve_ivp(
            flow,
            (t_curr, tf),
            y_curr,
            events=ground_event,
            dense_output=True,
            max_step=0.03,
            rtol=1e-8,
            atol=1e-10,
        )

        t_segments.append(sol.t)
        y_segments.append(sol.y.T)

        if sol.t_events[0].size == 0:
            break

        te = float(sol.t_events[0][0])
        ye_minus = sol.sol(te)
        impact_times.append(te)

        if abs(ye_minus[1]) < min_bounce_speed:
            # Stop after final small impact.
            t_segments.append(np.array([te]))
            y_segments.append(np.array([[0.0, 0.0]]))
            break

        y_plus = jump_map(ye_minus)
        t_curr = te
        # Restart slightly after the event to avoid immediate re-detection.
        y_curr = y_plus + np.array([0.0, 1e-12])

    return t_segments, y_segments, impact_times


if __name__ == "__main__":
    ts, ys, impacts = simulate_hybrid_bouncing_ball()

    t_plot = np.concatenate(ts)
    y_plot = np.vstack(ys)

    plt.figure(figsize=(8, 4))
    plt.plot(t_plot, y_plot[:, 0], label="height h(t)")
    for te in impacts:
        plt.axvline(te, linestyle="--", linewidth=0.8)
    plt.xlabel("Time (s)")
    plt.ylabel("Height (m)")
    plt.title("Hybrid Simulation: Bouncing Ball with Events")
    plt.grid(True)
    plt.legend()
    plt.tight_layout()
    plt.show()

    print("Impact times:", impacts)
